pkg/cloudprovider/providers/aws/aws.go

```diff
@@ -1777,36 +1791,60 @@ func (c *Cloud) DiskIsAttached(diskName KubernetesVolumeID, nodeName types.NodeN
 	return false, nil
 }
 
-func (c *Cloud) DisksAreAttached(diskNames []KubernetesVolumeID, nodeName types.NodeName) (map[KubernetesVolumeID]bool, error) {
-	idToDiskName := make(map[awsVolumeID]KubernetesVolumeID)
-	attached := make(map[KubernetesVolumeID]bool)
-	for _, diskName := range diskNames {
-		volumeID, err := diskName.mapToAWSVolumeID()
-		if err != nil {
-			return nil, fmt.Errorf("error mapping volume spec %q to aws id: %v", diskName, err)
-		}
-		idToDiskName[volumeID] = diskName
-		attached[diskName] = false
+func (c *Cloud) CheckClusterDisks(nodeDisks map[types.NodeName][]KubernetesVolumeID) (map[types.NodeName]map[KubernetesVolumeID]bool, error) {
+	attached := make(map[types.NodeName]map[KubernetesVolumeID]bool)
+
+	if len(nodeDisks) == 0 {
+		return attached, nil
 	}
-	_, instance, err := c.getFullInstance(nodeName)
+
+	dnsNameSlice := []string{}
+	for nodeName := range nodeDisks {
+		dnsNameSlice = append(dnsNameSlice, mapNodeNameToPrivateDNSName(nodeName))
+	}
+
+	awsInstances, err := c.getInstancesByNodeNames(dnsNameSlice)
 	if err != nil {
-		if err == cloudprovider.InstanceNotFound {
+		return nil, err
+	}
+
+	awsInstanceMap := make(map[types.NodeName]*ec2.Instance)
+	for _, awsInstance := range awsInstances {
+		awsInstanceMap[mapInstanceToNodeName(awsInstance)] = awsInstance
+	}
+
+	// Note that we check that the volume is attached to the correct node, not that it is attached to _a_ node
+	for nodeName, diskNames := range nodeDisks {
+		for _, diskName := range diskNames {
+			setNodeDisk(attached, diskName, nodeName, false)
+		}
+
+		awsInstance := awsInstanceMap[nodeName]
+		if awsInstance == nil {
 			// If instance no longer exists, safe to assume volume is not attached.
 			glog.Warningf(
 				"Node %q does not exist. DisksAreAttached will assume disks %v are not attached to it.",
 				nodeName,
 				diskNames)
-			return attached, nil
+			continue
 		}
 
-		return attached, err
-	}
-	for _, blockDevice := range instance.BlockDeviceMappings {
-		volumeID := awsVolumeID(aws.StringValue(blockDevice.Ebs.VolumeId))
-		diskName, found := idToDiskName[volumeID]
-		if found {
-			// Disk is still attached to node
-			attached[diskName] = true
+		idToDiskName := make(map[awsVolumeID]KubernetesVolumeID)
+		for _, diskName := range diskNames {
```

Comment:
it is bit tricky to roll this into same loop. We want by default disks to be marked as `detached`, that is what loop at line#1817 does. But if the node has disappeared from AWS, the we do `continue` and move to next node. But all disks on that node are considered detached.

If we combine the loops, then we wouldn't be able to bail out to next `continue` if node doesn't exist in AWS. 

It is of course possible to rearrange the code so that, we don't depend on the continue line:

```
		awsInstance := awsInstanceMap[nodeName]
		if awsInstance == nil {
			// If instance no longer exists, safe to assume volume is not attached.
			glog.Warningf(
				"Node %q does not exist. DisksAreAttached will assume disks %v are not attached to it.",
				nodeName,
				diskNames)
			continue
		}
```

But I feel we might be okay. @jingxu97 what you think?